### Bikesharing Assignment ML1

### import libraries

In [1]:

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt




# Data Reading

In [2]:
day_dataset= pd.read_csv('day.csv')
day_df=pd.DataFrame(day_dataset)
print(day_df)

     instant      dteday  season  yr  mnth  holiday  weekday  workingday  \
0          1  01-01-2018       1   0     1        0        6           0   
1          2  02-01-2018       1   0     1        0        0           0   
2          3  03-01-2018       1   0     1        0        1           1   
3          4  04-01-2018       1   0     1        0        2           1   
4          5  05-01-2018       1   0     1        0        3           1   
..       ...         ...     ...  ..   ...      ...      ...         ...   
725      726  27-12-2019       1   1    12        0        4           1   
726      727  28-12-2019       1   1    12        0        5           1   
727      728  29-12-2019       1   1    12        0        6           0   
728      729  30-12-2019       1   1    12        0        0           0   
729      730  31-12-2019       1   1    12        0        1           1   

     weathersit       temp     atemp      hum  windspeed  casual  registered  \
0      


#### converted csv dataset to Dataframe
- Its convenient to understand Data in   Table Format
- Lets get some standard information 


In [3]:
day_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 730 entries, 0 to 729
Data columns (total 16 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   instant     730 non-null    int64  
 1   dteday      730 non-null    object 
 2   season      730 non-null    int64  
 3   yr          730 non-null    int64  
 4   mnth        730 non-null    int64  
 5   holiday     730 non-null    int64  
 6   weekday     730 non-null    int64  
 7   workingday  730 non-null    int64  
 8   weathersit  730 non-null    int64  
 9   temp        730 non-null    float64
 10  atemp       730 non-null    float64
 11  hum         730 non-null    float64
 12  windspeed   730 non-null    float64
 13  casual      730 non-null    int64  
 14  registered  730 non-null    int64  
 15  cnt         730 non-null    int64  
dtypes: float64(4), int64(11), object(1)
memory usage: 88.5+ KB



#### Insights about data we got:
- Different types of data present 
- Column Names 
- Which column contains Null values
  with count
- No of data entries
#### Identification of categorical and continuous variables looking at datatypes

**Data types present**: 
 - int :continuous variable 
 - float:continuous variable
 - Object: categorical variable 

##### Now if we need stastical Insights about Data such as
**Mean,Median,Mode,Min ,Max that too 
for every column**


In [4]:
day_df.describe()

instant      season          yr        mnth     holiday     weekday  \
count  730.000000  730.000000  730.000000  730.000000  730.000000  730.000000   
mean   365.500000    2.498630    0.500000    6.526027    0.028767    2.997260   
std    210.877136    1.110184    0.500343    3.450215    0.167266    2.006161   
min      1.000000    1.000000    0.000000    1.000000    0.000000    0.000000   
25%    183.250000    2.000000    0.000000    4.000000    0.000000    1.000000   
50%    365.500000    3.000000    0.500000    7.000000    0.000000    3.000000   
75%    547.750000    3.000000    1.000000   10.000000    0.000000    5.000000   
max    730.000000    4.000000    1.000000   12.000000    1.000000    6.000000   

       workingday  weathersit        temp       atemp         hum   windspeed  \
count  730.000000  730.000000  730.000000  730.000000  730.000000  730.000000   
mean     0.683562    1.394521   20.319259   23.726322   62.765175   12.763620   
std      0.465405    0.544807    7.506729    8.150308   14.237589    5.195841   
min      0.000000    1.000000    2.424346    3.953480    0.000000    1.500244   
25%      0.000000    1.000000   13.811885   16.889713   52.000000    9.041650   
50%      1.000000    1.000000   20.465826   24.368225   62.625000   12.125325   
75%      1.000000    2.000000   26.880615   30.445775   72.989575   15.625589   
max      1.000000    3.000000   35.328347   42.044800   97.250000   34.000021   

            casual   registered          cnt  
count   730.000000   730.000000   730.000000  
mean    849.249315  3658.757534  4508.006849  
std     686.479875  1559.758728  1936.011647  
min       2.000000    20.000000    22.000000  
25%     316.250000  2502.250000  3169.750000  
50%     717.000000  3664.500000  4548.500000  
75%    1096.500000  4783.250000  5966.000000  
max    3410.000000  6946.000000  8714.000000

#### Structure of data such as :
- Number Of Rows,Columns,Dimensions

In [5]:
day_df.shape

(730, 16)

**No of Rows:730 , No of columns:16**

**Duplicate Rows Check & Missing value check**

In [6]:
day_df.duplicated().sum()
print(day_df.isna().sum())

instant       0
dteday        0
season        0
yr            0
mnth          0
holiday       0
weekday       0
workingday    0
weathersit    0
temp          0
atemp         0
hum           0
windspeed     0
casual        0
registered    0
cnt           0
dtype: int64


#### No missing values present

##### Derived new Matrix Day_name from column dteday

In [7]:
day_df['Day_name']=day_df['dteday']

In [8]:
day_df

instant      dteday  season  yr  mnth  holiday  weekday  workingday  \
0          1  01-01-2018       1   0     1        0        6           0   
1          2  02-01-2018       1   0     1        0        0           0   
2          3  03-01-2018       1   0     1        0        1           1   
3          4  04-01-2018       1   0     1        0        2           1   
4          5  05-01-2018       1   0     1        0        3           1   
..       ...         ...     ...  ..   ...      ...      ...         ...   
725      726  27-12-2019       1   1    12        0        4           1   
726      727  28-12-2019       1   1    12        0        5           1   
727      728  29-12-2019       1   1    12        0        6           0   
728      729  30-12-2019       1   1    12        0        0           0   
729      730  31-12-2019       1   1    12        0        1           1   

     weathersit       temp     atemp      hum  windspeed  casual  registered  \
0             2  14.110847  18.18125  80.5833  10.749882     331         654   
1             2  14.902598  17.68695  69.6087  16.652113     131         670   
2             1   8.050924   9.47025  43.7273  16.636703     120        1229   
3             1   8.200000  10.60610  59.0435  10.739832     108        1454   
4             1   9.305237  11.46350  43.6957  12.522300      82        1518   
..          ...        ...       ...      ...        ...     ...         ...   
725           2  10.420847  11.33210  65.2917  23.458911     247        1867   
726           2  10.386653  12.75230  59.0000  10.416557     644        2451   
727           2  10.386653  12.12000  75.2917   8.333661     159        1182   
728           1  10.489153  11.58500  48.3333  23.500518     364        1432   
729           2   8.849153  11.17435  57.7500  10.374682     439        2290   

      cnt    Day_name  
0     985  01-01-2018  
1     801  02-01-2018  
2    1349  03-01-2018  
3    1562  04-01-2018  
4    1600  05-01-2018  
..    ...         ...  
725  2114  27-12-2019  
726  3095  28-12-2019  
727  1341  29-12-2019  
728  1796  30-12-2019  
729  2729  31-12-2019  

[730 rows x 17 columns]

**New Day_name column created,we already
  have yr,month column 
  so dte_day column not required.**
  - We can safely drop dte_day column

In [9]:
day_df=day_df.drop('dteday',axis=1)
day_df

instant  season  yr  mnth  holiday  weekday  workingday  weathersit  \
0          1       1   0     1        0        6           0           2   
1          2       1   0     1        0        0           0           2   
2          3       1   0     1        0        1           1           1   
3          4       1   0     1        0        2           1           1   
4          5       1   0     1        0        3           1           1   
..       ...     ...  ..   ...      ...      ...         ...         ...   
725      726       1   1    12        0        4           1           2   
726      727       1   1    12        0        5           1           2   
727      728       1   1    12        0        6           0           2   
728      729       1   1    12        0        0           0           1   
729      730       1   1    12        0        1           1           2   

          temp     atemp      hum  windspeed  casual  registered   cnt  \
0    14.110847  18.18125  80.5833  10.749882     331         654   985   
1    14.902598  17.68695  69.6087  16.652113     131         670   801   
2     8.050924   9.47025  43.7273  16.636703     120        1229  1349   
3     8.200000  10.60610  59.0435  10.739832     108        1454  1562   
4     9.305237  11.46350  43.6957  12.522300      82        1518  1600   
..         ...       ...      ...        ...     ...         ...   ...   
725  10.420847  11.33210  65.2917  23.458911     247        1867  2114   
726  10.386653  12.75230  59.0000  10.416557     644        2451  3095   
727  10.386653  12.12000  75.2917   8.333661     159        1182  1341   
728  10.489153  11.58500  48.3333  23.500518     364        1432  1796   
729   8.849153  11.17435  57.7500  10.374682     439        2290  2729   

       Day_name  
0    01-01-2018  
1    02-01-2018  
2    03-01-2018  
3    04-01-2018  
4    05-01-2018  
..          ...  
725  27-12-2019  
726  28-12-2019  
727  29-12-2019  
728  30-12-2019  
729  31-12-2019  

[730 rows x 16 columns]

- Irrelevant column dte_day removed without any dataloss

### Data preparation

#### yr month already exist ,now we want
#### data to be more meaningful
### Data Transformation 
*I want Day column to have values like*
**Monday,Tuesday,Wednesday,Thursday,Friday,Saturday,Sunday
corresponding to value it has like 
01,02,03,04,05,06,07**

In [10]:
 days=[el.split('-')[0] for el in day_df['Day_name']]

In [11]:
 names=[]
for el in day_df['Day_name']:
    names.append(
    pd.to_datetime(el,dayfirst=True).strftime('%A'))
     

In [12]:
No_Of_Entries=len(day_df)
No_Of_Weeks= round(No_Of_Entries/7)
No_Of_Months=round(No_Of_Weeks/4)
Years_count=round(No_Of_Months/12)
print('No_Of_Months:',No_Of_Months)
print('No_Of_Weeks:',No_Of_Weeks)
print('No_Of_Years:',Years_count)


No_Of_Months: 26
No_Of_Weeks: 104
No_Of_Years: 2


In [13]:
day_df['Day_name']=names
print(day_df['Day_name'])

0         Monday
1        Tuesday
2      Wednesday
3       Thursday
4         Friday
         ...    
725       Friday
726     Saturday
727       Sunday
728       Monday
729      Tuesday
Name: Day_name, Length: 730, dtype: object


### Data Visualization

##### Barplot 

In [14]:

    x=day_df['Day_name']
    y=day_df['cnt']
    plt.bar(x,y)
    plt.show()